File: apps/mobile/src/native/session.ts
import { PermissionReply, type ThreadSnapshot } from 'trex-core';
import { useCallback, useEffect, useRef, useState } from 'react';

import { toChatImage, type Attachment } from './attachments';
import { useClient } from './client';
import { describeError, isUnknownSession } from './errors';
import { useSessionGone } from './session-presence';
import {
  EMPTY_THREAD,
  parseThread,
  type AskQuestion,
  type PermissionSuggestion,
  type QuestionAnswers,
  type Thread,
} from './thread';

export type SessionView = {
  thread: Thread;
  /** True until the first snapshot lands — the core pushes one even for an
   * empty session, so this is a real signal rather than a guess. */
  loading: boolean;
  /** A failed action (send/resolve/cancel), shown inline and dismissible. */
  error?: string;
  /**
   * The desktop has closed this session. Terminal — nothing will arrive and no
   * action will succeed — so the screen stops offering to drive it rather than
   * letting every control fail one by one.
   */
  closed: boolean;
  /** Resolves `false` if the send failed, so the composer can keep the text. */
  send: (text: string, images?: Attachment[]) => Promise<boolean>;
  /** Guide a turn that is already running, instead of queueing a new prompt. */
  steer: (text: string) => Promise<boolean>;
  cancel: () => Promise<boolean>;
  allow: (requestId: string, toolInput: unknown) => Promise<boolean>;
  allowWith: (
    requestId: string,
    toolInput: unknown,
    suggestion: PermissionSuggestion
  ) => Promise<boolean>;
  deny: (requestId: string, message: string) => Promise<boolean>;
  answer: (
    requestId: string,
    questions: AskQuestion[],
    answers: QuestionAnswers
  ) => Promise<boolean>;
  /**
   * Rewind to a user turn: drop it and everything after it.
   *
   * `ordinal` counts **user** entries only, matching the wire. Nothing is
   * applied locally — the truncation arrives as a `Rewound` event that folds
   * through the same subscription, so the phone and the desktop agree by
   * construction rather than by two implementations happening to match.
   */
  rewind: (ordinal: number) => Promise<boolean>;
  dismissError: () => void;
  /** Surface a failure that happened outside an RPC (picking an image, say)
   * through the same inline banner, so the screen has one error channel. */
  reportError: (message: string) => void;
};

/**
 * Subscribe to one session and expose its folded transcript plus the actions
 * that drive it.
 *
 * The thread arrives already folded from the Rust core, so this hook only
 * parses and stores it — there is no reduction here, deliberately: the fold is
 * `agent-core`'s, shared with the desktop.
 *
 * Note the core has no `unsubscribe`: leaving a session leaves its subscription
 * registered until the client disconnects, and it keeps folding in the
 * background. That is why the sink below guards on `live` — a subscription
 * outliving its screen would otherwise push into an unmounted component.
 * Re-entering a session is *not* free, though: `subscribe` re-fetches the
 * backlog from seq 0 and folds a fresh thread, so the cost scales with the
 * transcript rather than being a cheap re-attach.
 */
export function useSession(sessionId: string): SessionView {
  const client = useClient((s) => s.client);
  const [thread, setThread] = useState<Thread>(EMPTY_THREAD);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string>();
  // The two liveness signals `useSessionGone` weighs. Kept apart from `error`:
  // a session the desktop has closed is a state the screen renders, not a
  // failure the user can dismiss and retry past.
  const [unknownSession, setUnknownSession] = useState(false);
  const [subscribed, setSubscribed] = useState(false);
  // Actions read the client through a ref so a reconnect (which swaps the client)
  // does not have to re-create every callback the composer holds.
  const clientRef = useRef(client);
  clientRef.current = client;

  useEffect(() => {
    if (!client) return;
    let live = true;
    // The highest fold cursor rendered so far. Snapshots are whole-state, so an
    // older one arriving late (a re-subscribe racing the live dispatcher after a
    // reconnect) would visibly rewind the transcript. Dropping anything below the
    // high-water mark makes the rendered thread monotonic.
    let renderedSeq = -1;
    setLoading(true);
    // A reconnect re-runs this against a fresh client; neither the previous
    // attempt's verdict nor its confirmation may outlive it — the desktop may
    // have closed the session in between, or reopened one it had closed.
    setUnknownSession(false);
    setSubscribed(false);

    const sink = {
      onThread: (snapshot: ThreadSnapshot) => {
        if (!live) return;
        const seq = Number(snapshot.seq);
        if (seq < renderedSeq) return;
        renderedSeq = seq;
        // A snapshot that will not parse must not blank the transcript: keep
        // rendering the last good one rather than dropping to an empty screen.
        try {
          setThread(parseThread(snapshot.threadJson));
          setLoading(false);
        } catch (e) {
          setError(describeError(e));
        }
      },
    };

    client
      .subscribe(sessionId, sink)
      .then(() => {
        // The host answered about this session, so it had it. That is what
        // later lets its absence from the session list be read as a close.
        if (live) setSubscribed(true);
      })
      .catch((e: unknown) => {
        if (!live) return;
        setLoading(false);
        // "No such session" is not an error to show: the desktop closed the tab,
        // so the screen says that instead of quoting the wire back at the user.
        if (isUnknownSession(e)) setUnknownSession(true);
        else setError(describeError(e));
      });

    return () => {
      live = false;
    };
  }, [client, sessionId]);

  /**
   * Run an action, surfacing failure inline rather than throwing into render.
   * Reports whether it succeeded so a caller holding user input (the composer)
   * can decide whether it is safe to discard it.
   */
  const run = useCallback(
    async (action: (c: NonNullable<typeof client>) => Promise<unknown>): Promise<boolean> => {
      const current = clientRef.current;
      if (!current) {
        setError('Not connected to the desktop.');
        return false;
      }
      try {
        await action(current);
        return true;
      } catch (e) {
        setError(describeError(e));
        return false;
      }
    },
    []
  );

  const send = useCallback(
    (text: string, images: Attachment[] = []) =>
      run((c) => c.sendPrompt(sessionId, text, images.map(toChatImage))),
    [run, sessionId]
  );

  const steer = useCallback(
    (text: string) => run((c) => c.steer(sessionId, text)),
    [run, sessionId]
  );

  const cancel = useCallback(() => run((c) => c.cancel(sessionId)), [run, sessionId]);

  const rewind = useCallback(
    // `includeFiles: false` — the phone never restores the working tree. That
    // discards uncommitted work belonging to whoever is at the desktop, which
    // is not a thing to trigger from a pocket.
    (ordinal: number) => run((c) => c.rewindSession(sessionId, ordinal, false)),
    [run, sessionId]
  );

  const allow = useCallback(
    (requestId: string, toolInput: unknown) =>
      run((c) =>
        c.resolvePermission(
          sessionId,
          requestId,
          // The allow MUST echo the tool's input back: the CLI treats an allow
          // without it as malformed and denies the tool instead. The input we
          // echo is the one already on the tool call this request belongs to.
          new PermissionReply.Allow({ updatedInputJson: JSON.stringify(toolInput ?? null) })
        )
      ),
    [run, sessionId]
  );

  const allowWith = useCallback(
    (requestId: string, toolInput: unknown, suggestion: PermissionSuggestion) =>
      run((c) =>
        c.resolvePermission(
          sessionId,
          requestId,
          // The suggestion is quoted straight back from the card that displayed
          // it. Its `raw` is opaque here — only the agent that offered it knows
          // what it means — so echoing rather than rebuilding is what keeps the
          // phone from inventing a grant the agent never proposed.
          new PermissionReply.AllowWithSuggestion({
            updatedInputJson: JSON.stringify(toolInput ?? null),
            suggestionJson: JSON.stringify(suggestion),
          })
        )
      ),
    [run, sessionId]
  );

  const deny = useCallback(
    (requestId: string, message: string) =>
      run((c) => c.resolvePermission(sessionId, requestId, new PermissionReply.Deny({ message }))),
    [run, sessionId]
  );

  const answer = useCallback(
    (requestId: string, questions: AskQuestion[], answers: QuestionAnswers) =>
      run((c) =>
        // Both payloads cross as JSON: the questions are quoted straight back
        // from the snapshot that produced this card, so the core sees the exact
        // shapes the fold emitted rather than a re-derived approximation.
        c.answerQuestion(
          sessionId,
          requestId,
          JSON.stringify(questions),
          JSON.stringify(answers)
        )
      ),
    [run, sessionId]
  );

  const dismissError = useCallback(() => setError(undefined), []);
  const reportError = useCallback((message: string) => setError(message), []);

  const closed = useSessionGone(sessionId, { subscribed, unknownSession });

  return {
    thread,
    loading,
    error,
    closed,
    send,
    steer,
    cancel,
    allow,
    allowWith,
    deny,
    answer,
    rewind,
    dismissError,
    reportError,
  };
}

File: crates/remote-session/src/session.rs
//! The client session: the Register/Connect/AuthProve handshake and the one-shot
//! RPCs, driven over the `remote-proto` [`Transport`] seam. Pure Rust — the iroh
//! transport is injected as one `Transport` impl, and the in-memory loopback
//! drives the full-loop test against the real host dispatcher.
//!
//! Every RPC rides the concurrent [`demux`](crate::demux) pump, so a client can
//! issue requests *while* subscribed to a live session on the same connection: the
//! pump routes pushed `HostEvent`s to the event stream ([`Self::take_events`]) and
//! each reply to its waiting caller. The reconnect state machine is a later slice.

mod git;
mod handshake;
mod subscribe;
mod terminals;

use std::sync::{Arc, Mutex};

use futures::channel::oneshot;
use trex_agent_core::thread::{AskQuestion, ChatImage, PermissionDecision, QuestionAnswers};
use trex_remote_proto::messages::{
    AnswerQuestionReq, CheckRunWire, ForgeItemDetailWire, ForgeItemKindWire, ForgeItemWire,
    ForgeStateWire, ProjectSummaryWire, RecurrenceV2Wire, RecurrenceWire, ResolvePermissionReq,
    ScheduleRunWire, ScheduleV2Wire, ScheduleWire, SendPromptReq, SessionInfoWire, SessionSummary, SessionTranscriptWire,
};
use trex_remote_proto::proto::{Request, Response, RpcError, SessionChoices};
use trex_remote_proto::{HostEvent, Transport};

use crate::demux::{Demux, DemuxPump, EventStream, SessionsStream, TerminalStream, demux};
use crate::error::SessionError;
use crate::signer::ClientSigner;

type Result<T> = std::result::Result<T, SessionError>;

/// One client's connection to a host, over an abstract [`Transport`]. Owns the
/// app-signing identity, the demux RPC handle, and caches the reconnect token.
pub struct RemoteSession {
    demux: Arc<Demux>,
    signer: ClientSigner,
    /// The reconnect credential the host issues on Register/Connect. Cached in
    /// memory only — `mobile-core` may persist it, but losing it just forces the
    /// slower Ed25519 challenge on the next `connect`.
    token: Mutex<Option<String>>,
    /// The read-loop pump, taken once by the owner to drive (spawned in prod,
    /// joined in tests). Every RPC is dead in the water until it runs.
    pump: Mutex<Option<DemuxPump>>,
    /// The live event stream, taken once by the owner to consume.
    events: Mutex<Option<EventStream>>,
    /// The live terminal stream, taken once by the owner to consume. Separate
    /// from `events` because a client can watch a terminal without subscribing
    /// to any agent session, and vice versa.
    terminals: Mutex<Option<TerminalStream>>,
    /// The pushed session-list stream, taken once by the owner. Each item is a full
    /// per-device snapshot that replaces the client's session list wholesale.
    sessions: Mutex<Option<SessionsStream>>,
    /// Dropping this stops the pump — so the connection tears down when the
    /// session is dropped, no explicit close needed.
    _shutdown: oneshot::Sender<()>,
}

impl RemoteSession {
    pub fn new(transport: Arc<dyn Transport>, signer: ClientSigner) -> Self {
        let (handle, pump, events, terminals, sessions, shutdown) = demux(transport);
        Self {
            demux: handle,
            signer,
            token: Mutex::new(None),
            pump: Mutex::new(Some(pump)),
            events: Mutex::new(Some(events)),
            terminals: Mutex::new(Some(terminals)),
            sessions: Mutex::new(Some(sessions)),
            _shutdown: shutdown,
        }
    }

    /// Take the pushed session-list stream — once. Each item is a full snapshot the
    /// owner uses to replace its session list.
    pub fn take_sessions(&self) -> Option<SessionsStream> {
        self.sessions.lock().unwrap().take()
    }

    /// Subscribe to the live session list. Returns the current snapshot immediately;
    /// thereafter the host pushes a fresh snapshot on every change onto the stream
    /// from [`Self::take_sessions`]. Idempotent — a repeat subscribe re-snapshots.
    pub async fn subscribe_sessions(&self) -> Result<Vec<SessionSummary>> {
        // The immediate reply is a plain `Sessions` snapshot (routed to the RPC
        // slot); subsequent changes arrive as pushed `SessionsChanged` frames on the
        // stream from `take_sessions`.
        match self.call(Request::SubscribeSessions).await? {
            Response::Sessions(rows) => Ok(rows),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "Sessions" }),
        }
    }

    /// Take the read-loop pump to drive it — once. Spawn `pump.run()` onto an
    /// executor (prod) or join it (tests); RPCs only resolve while it runs.
    pub fn take_pump(&self) -> Option<DemuxPump> {
        self.pump.lock().unwrap().take()
    }

    /// Take the live event stream — once. Each pushed `HostEvent` is folded by a
    /// [`SessionSubscription`](crate::SessionSubscription).
    pub fn take_terminals(&self) -> Option<TerminalStream> {
        self.terminals.lock().unwrap().take()
    }

    /// Take the live event stream — once. Each pushed `HostEvent` is folded by a
    /// [`SessionSubscription`](crate::SessionSubscription).
    pub fn take_events(&self) -> Option<EventStream> {
        self.events.lock().unwrap().take()
    }

    /// The app-signing public key the host records for this device.
    pub fn public_key(&self) -> [u8; 32] {
        self.signer.public_key()
    }

    /// The cached reconnect token, if any (for the caller to persist).
    pub fn session_token(&self) -> Option<String> {
        self.token.lock().unwrap().clone()
    }

    /// Seed a persisted reconnect token so the next [`Self::connect`] can take the
    /// fast path.
    pub fn set_session_token(&self, token: Option<String>) {
        *self.token.lock().unwrap() = token;
    }

    // ---- one-shot session RPCs (the handshake lives in `handshake.rs`) ----

    /// Every session the host exposes to this device.
    pub async fn list_sessions(&self) -> Result<Vec<SessionSummary>> {
        match self.call(Request::ListSessions).await? {
            Response::Sessions(sessions) => Ok(sessions),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "Sessions" }),
        }
    }

    /// One session's detail + resume cursor.
    pub async fn session_info(&self, session_id: &str) -> Result<SessionInfoWire> {
        let req = Request::GetSessionInfo { session_id: session_id.to_string() };
        match self.call(req).await? {
            Response::SessionInfo(info) => Ok(info),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "SessionInfo" }),
        }
    }

    /// The session's authoritative folded-transcript snapshot — full history plus
    /// the `seq` it reflects. Fetched when opening a session so it renders its
    /// history immediately (including a restart-restored transcript that never
    /// entered the live ring); the caller then subscribes from `seq` to extend it.
    pub async fn fetch_transcript(&self, session_id: &str) -> Result<SessionTranscriptWire> {
        let req = Request::FetchTranscript { session_id: session_id.to_string() };
        match self.call(req).await? {
            Response::SessionTranscript(t) => Ok(t),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "SessionTranscript" }),
        }
    }

    /// The desktop's projects, offered as new-session targets so the phone can
    /// start a session in one by its path instead of typing it. May be empty.
    pub async fn list_projects(&self) -> Result<Vec<ProjectSummaryWire>> {
        match self.call(Request::ListProjects).await? {
            Response::Projects(rows) => Ok(rows),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "Projects" }),
        }
    }

    /// Drop this device's enrollment on the host, so forgetting the desktop here
    /// also clears this device from the desktop's paired-devices list.
    ///
    /// One-way and immediate: every later RPC on this connection is refused, so
    /// callers unpair as the last thing they do before tearing the link down.
    pub async fn unpair(&self) -> Result<()> {
        self.expect_ack(Request::Unpair).await
    }

    /// Send a user prompt into a session, starting a turn. `corr_id` lets the
    /// caller match the eventual echoed turn in the event stream.
    pub async fn send_prompt(
        &self,
        session_id: &str,
        text: &str,
        images: &[ChatImage],
        corr_id: u64,
    ) -> Result<()> {
        let req = Request::SendPrompt(SendPromptReq {
            session_id: session_id.to_string(),
            text: text.to_string(),
            images: images.to_vec(),
            corr_id,
        });
        self.expect_ack(req).await
    }

    /// Answer an outstanding permission request. `Ok(true)` = this call decided it;
    /// `Ok(false)` = it was already decided (idempotent — still a success).
    pub async fn resolve_permission(
        &self,
        session_id: &str,
        request_id: &str,
        decision: &PermissionDecision,
    ) -> Result<bool> {
        let payload = ResolvePermissionReq::new(session_id, request_id, decision)
            .map_err(|e| SessionError::Wire(e.to_string()))?;
        match self.call(Request::ResolvePermission(payload)).await? {
            Response::Ack => Ok(true),
            Response::Error(RpcError::AlreadyDecided) => Ok(false),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "Ack" }),
        }
    }

    /// Answer an outstanding `AskUserQuestion`. Same idempotency contract as
    /// [`Self::resolve_permission`]: `Ok(false)` means someone answered first.
    pub async fn answer_question(
        &self,
        session_id: &str,
        request_id: &str,
        questions: &[AskQuestion],
        answers: &QuestionAnswers,
    ) -> Result<bool> {
        let payload = AnswerQuestionReq {
            session_id: session_id.to_string(),
            request_id: request_id.to_string(),
            questions: questions.to_vec(),
            answers: answers.clone(),
        };
        match self.call(Request::AnswerQuestion(payload)).await? {
            Response::Ack => Ok(true),
            Response::Error(RpcError::AlreadyDecided) => Ok(false),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "Ack" }),
        }
    }

    /// Steer a mid-turn agent with extra guidance.
    pub async fn steer(&self, session_id: &str, text: &str) -> Result<()> {
        let req = Request::Steer { session_id: session_id.to_string(), text: text.to_string() };
        self.expect_ack(req).await
    }

    /// Cancel the session's in-flight turn.
    pub async fn cancel(&self, session_id: &str) -> Result<()> {
        self.expect_ack(Request::Cancel { session_id: session_id.to_string() }).await
    }

    /// The models and permission modes this session's backend offers.
    ///
    /// Empty lists mean "nothing to choose from" — a dynamic-catalog backend
    /// before its handshake completes, or an agent with no mode options — not a
    /// failure.
    pub async fn list_choices(&self, session_id: &str) -> Result<SessionChoices> {
        let req = Request::ListChoices { session_id: session_id.to_string() };
        match self.call(req).await? {
            Response::Choices(choices) => Ok(choices),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "Choices" }),
        }
    }

    /// Switch the session's model.
    ///
    /// Errors when the backend fixes its model at spawn time; the host answers
    /// with a message saying so rather than accepting silently, so this is worth
    /// surfacing to the user instead of swallowing.
    pub async fn set_model(&self, session_id: &str, model: &str) -> Result<()> {
        let req = Request::SetModel {
            session_id: session_id.to_string(),
            model: model.to_string(),
        };
        self.expect_ack(req).await
    }

    /// Switch the session's permission mode. Same fix-at-spawn caveat as
    /// [`Self::set_model`].
    pub async fn set_permission_mode(&self, session_id: &str, mode: &str) -> Result<()> {
        let req = Request::SetPermissionMode {
            session_id: session_id.to_string(),
            mode: mode.to_string(),
        };
        self.expect_ack(req).await
    }

    /// Start a new agent session on the desktop, returning its id.
    ///
    /// The id comes back rather than being discovered by re-listing, so the
    /// caller can subscribe to the new session immediately instead of polling
    /// and guessing which row appeared.
    pub async fn create_session(&self, cwd: &str, agent_id: Option<&str>) -> Result<String> {
        let req = Request::CreateSession {
            cwd: cwd.to_string(),
            agent_id: agent_id.map(str::to_string),
        };
        match self.call(req).await? {
            Response::SessionCreated { session_id } => Ok(session_id),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "SessionCreated" }),
        }
    }

    /// Rewind a session to an earlier turn, dropping that turn and everything
    /// after it.
    ///
    /// **Destructive and not undoable from here.** Returns once the host has
    /// accepted the rewind, not once it has finished: the truncation arrives on
    /// the event stream as `ThreadEvent::Rewound`, which is also how a
    /// desktop-initiated rewind reaches this client, so there is one code path
    /// applying it rather than two.
    ///
    /// `include_files` additionally restores the working tree to the turn's
    /// checkpoint, discarding uncommitted work. The host may refuse it while
    /// still performing the conversation rewind's validation; a caller should
    /// treat that refusal as normal rather than retrying.
    pub async fn rewind_session(
        &self,
        session_id: &str,
        ordinal: u32,
        include_files: bool,
    ) -> Result<()> {
        let req = Request::RewindSession {
            session_id: session_id.to_string(),
            ordinal,
            include_files,
        };
        self.expect_ack(req).await
    }

    /// Issues or pull requests for the session's repository.
    ///
    /// **An empty list is a normal answer**, not a failure: a repo hosted
    /// nowhere relevant, a forge CLI that is absent or signed out, or simply no
    /// matching items all resolve to empty. The host cannot tell those apart, so
    /// a caller must render "nothing here" rather than inventing a reason.
    pub async fn list_forge_items(
        &self,
        session_id: &str,
        kind: ForgeItemKindWire,
        state: ForgeStateWire,
        mine: bool,
    ) -> Result<Vec<ForgeItemWire>> {
        let req = Request::ListForgeItems {
            session_id: session_id.to_string(),
            kind,
            state,
            mine,
        };
        match self.call(req).await? {
            Response::ForgeItems(items) => Ok(items),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "ForgeItems" }),
        }
    }

    /// Body + author of one issue/PR. `None` when the forge CLI could not supply
    /// it — distinct from an item whose body is genuinely empty.
    pub async fn forge_item_detail(
        &self,
        session_id: &str,
        kind: ForgeItemKindWire,
        number: u64,
    ) -> Result<Option<ForgeItemDetailWire>> {
        let req = Request::GetForgeItemDetail {
            session_id: session_id.to_string(),
            kind,
            number,
        };
        match self.call(req).await? {
            Response::ForgeItemDetail(detail) => Ok(detail),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "ForgeItemDetail" }),
        }
    }

    /// CI check runs for the current branch's pull request. Empty is normal —
    /// no PR, no checks, or a forge with no pipeline mapping.
    pub async fn list_forge_checks(&self, session_id: &str) -> Result<Vec<CheckRunWire>> {
        let req = Request::ListForgeChecks { session_id: session_id.to_string() };
        match self.call(req).await? {
            Response::ForgeChecks(checks) => Ok(checks),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "ForgeChecks" }),
        }
    }

    /// Every schedule the desktop holds. **Empty is a normal answer** — a desktop
    /// with no schedules is the common case. Refused for a session-scoped device,
    /// which has no session to be narrowed to.
    pub async fn list_schedules(&self) -> Result<Vec<ScheduleWire>> {
        match self.call(Request::ListSchedules).await? {
            Response::Schedules(rows) => Ok(rows),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "Schedules" }),
        }
    }

    /// [`Self::list_schedules`] against a v23 host: the same rows, with cron
    /// expressions intact.
    ///
    /// Callers that can decode this should prefer it — the v18 reply substitutes
    /// a stand-in recurrence for a cron schedule, which is fine to display and
    /// wrong to reason about. Gate on
    /// [`SCHEDULE_CRON_MIN_VERSION`](trex_remote_proto::proto::SCHEDULE_CRON_MIN_VERSION)
    /// before calling: an older host cannot decode the request frame at all.
    pub async fn list_schedules_v2(&self) -> Result<Vec<ScheduleV2Wire>> {
        match self.call(Request::ListSchedulesV2).await? {
            Response::SchedulesV2(rows) => Ok(rows),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "SchedulesV2" }),
        }
    }

    /// Create a schedule, returning the stored row — its derived id and first
    /// next-fire come back so the caller can show it without re-listing.
    ///
    /// An invalid recurrence (an interval under the desktop's floor, an
    /// impossible time) is refused by the host rather than stored; the phone's
    /// pickers cannot express those, so this only bites a malformed caller.
    pub async fn create_schedule(
        &self,
        name: &str,
        cwd: &str,
        prompt: &str,
        agent_id: Option<&str>,
        recurrence: RecurrenceWire,
    ) -> Result<ScheduleWire> {
        let req = Request::CreateSchedule {
            name: name.to_string(),
            cwd: cwd.to_string(),
            prompt: prompt.to_string(),
            agent_id: agent_id.map(str::to_string),
            recurrence,
        };
        match self.call(req).await? {
            Response::ScheduleCreated(sched) => Ok(sched),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "ScheduleCreated" }),
        }
    }

    /// [`Self::create_schedule`] against a v23 host, whose recurrence may be a
    /// cron expression.
    ///
    /// The expression is validated host-side: one that will not parse as five
    /// fields, one that can never fire, and one tighter than the interval floor
    /// all come back as a `BadRequest` naming the rule broken.
    pub async fn create_schedule_v2(
        &self,
        name: &str,
        cwd: &str,
        prompt: &str,
        agent_id: Option<&str>,
        recurrence: RecurrenceV2Wire,
    ) -> Result<ScheduleV2Wire> {
        let req = Request::CreateScheduleV2 {
            name: name.to_string(),
            cwd: cwd.to_string(),
            prompt: prompt.to_string(),
            agent_id: agent_id.map(str::to_string),
            recurrence,
        };
        match self.call(req).await? {
            Response::ScheduleCreatedV2(sched) => Ok(sched),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "ScheduleCreatedV2" }),
        }
    }

    /// Delete a schedule. Idempotent — deleting one already gone is success.
    /// In-flight runs are unaffected; this stops future fires.
    pub async fn delete_schedule(&self, id: &str) -> Result<()> {
        self.expect_ack(Request::DeleteSchedule { id: id.to_string() }).await
    }

    /// Enable or disable a schedule without deleting it. Re-enabling recomputes
    /// the next fire from now — a schedule paused for a week does not wake owing a
    /// week of missed runs.
    pub async fn set_schedule_enabled(&self, id: &str, enabled: bool) -> Result<()> {
        self.expect_ack(Request::SetScheduleEnabled { id: id.to_string(), enabled }).await
    }

    /// A schedule's recent run history, most recent first, capped at `limit`.
    /// **Empty means it has never fired**, normal for a fresh schedule.
    pub async fn get_schedule_runs(
        &self,
        schedule_id: &str,
        limit: u32,
    ) -> Result<Vec<ScheduleRunWire>> {
        let req = Request::GetScheduleRuns { schedule_id: schedule_id.to_string(), limit };
        match self.call(req).await? {
            Response::ScheduleRuns(rows) => Ok(rows),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "ScheduleRuns" }),
        }
    }

    /// Transcribe a recorded voice clip with the desktop's speech engine.
    ///
    /// `audio_base64` is a standard-base64 WAV (16 kHz mono PCM16 is the phone's
    /// contract). An **empty transcript is a normal answer** — a silent clip, or
    /// one that held only filler — not an error; the caller inserts it as-is.
    pub async fn transcribe_audio(&self, audio_base64: &str, sample_rate: u32) -> Result<String> {
        let req = Request::TranscribeAudio { audio_base64: audio_base64.to_string(), sample_rate };
        match self.call(req).await? {
            Response::Transcript(text) => Ok(text),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "Transcript" }),
        }
    }

    /// Gap-fill: replay retained events after `after_seq` (the resync path when the
    /// live stream reports a `seq` jump).
    pub async fn events_since(&self, session_id: &str, after_seq: u64) -> Result<Vec<HostEvent>> {
        let req = Request::EventsSince { session_id: session_id.to_string(), after_seq };
        match self.call(req).await? {
            Response::Events(events) => Ok(events),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "Events" }),
        }
    }

    // ---- plumbing ----

    fn cache(&self, token: String) {
        *self.token.lock().unwrap() = Some(token);
    }

    /// A command whose only successful reply is `Ack`.
    async fn expect_ack(&self, req: Request) -> Result<()> {
        match self.call(req).await? {
            Response::Ack => Ok(()),
            Response::Error(e) => Err(SessionError::Rpc(e)),
            _ => Err(SessionError::Unexpected { expected: "Ack" }),
        }
    }

    /// Send one request and await its reply. Rides the [`demux`](crate::demux)
    /// pump, which routes pushed `Response::Event` frames off to the event stream
    /// and this reply back here — so an RPC is safe to issue even while a live
    /// subscription is streaming on the same connection.
    async fn call(&self, req: Request) -> Result<Response> {
        self.demux.call(req).await
    }
}
